### PR TITLE
Migrates PurchaserInfo

### PR DESCRIPTION
**Pull request by beylmk**

### CODE REVIEW

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+    
+     @return The purchase date for `entitlement`, `nil` if product never purchased
+     */
+    @objc public func purchaseDate(forEntitlement: String) -> Date? {
+        return entitlements[forEntitlement]?.latestPurchaseDate
+    }
+
+    // TODO why were we previously able to call with lowercase from swift?
+    // TODO after migration make this internal
+    @objc(JSONObject) public func jsonObject() -> [AnyHashable: Any] {
```
Comment (beylmk) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:174]: This was uppercase `JSONObject` in objc before, but was called from swift with lowercase `jsonObject`...does anyone know how that ever worked? In either case, this should be internal, so we can migrate to lowercase post-migration...

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:174]: I think it's just the obj-c -> swift naming translation

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:174]: https://github.com/apple/swift-evolution/blob/master/proposals/0005-objective-c-name-translation.md

> Lowercase values: The Swift API Design Guidelines have non-type declarations lowercased. Lowercase non-prefixed values whenever they are imported, including enumerators (whether they end up in Swift as enum cases or option sets) and any other properties/functions (e.g., a property named URLHandler will be lowercased to urlHandler).

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+//  PurchaserInfo.swift
+//  PurchasesCoreSwift
+//
+//  Created by Madeline Beyl on 7/9/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+
+@objc(RCPurchaserInfo) public class PurchaserInfo: NSObject {
```
Comment (beylmk) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:11]: I had trouble running jazzy on my M1...working on that, so for now consider comments not ready for review...

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 1 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:162]: are the keys not guaranteed to be strings? if they are, we should constrain it to `[String: Any]`, since it'll make stuff easier to compare and use, and introduce more safety

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        guard let other = object as? PurchaserInfo else {
+            return false
+        }
+
+        var selfJson = self.jsonObject()
+        selfJson.removeValue(forKey: "request_date")
+        var otherJson = other.jsonObject()
+        otherJson.removeValue(forKey: "request_date")
+
+        return NSDictionary(dictionary: selfJson).isEqual(to: otherJson)
```
[the remark this answers is not in the document]
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:176]: also, do we need these to be NSDictionaries?
would doing something `selfJson == otherJson` not work? is it because of the `AnyHashable` thing?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        description += "originalApplicationVersion=\(String(describing: self.originalApplicationVersion)),\n"
+        description += "latestExpirationDate=\(String(describing: self.latestExpirationDate)),\n"
+        description += "activeEntitlements=\(activeEntitlementsDescription),\n"
+        description += "activeSubscriptions=\(activeSubsDescription),\n"
+        description += "nonSubscriptionTransactions=\(self.nonSubscriptionTransactions),\n"
+        description += "requestDate=\(String(describing: self.requestDate)),\n"
+        description += "firstSeen=\(String(describing: self.firstSeen)),\n"
+        description += "originalAppUserId=\(self.originalAppUserId),\n"
+        description += "entitlements=\(allEntitlementsDescription),\n"
+        description += ">"
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:198]: this being swift, we could make it nicer now: 
```swift
return """
<\(String(describing: self): 
originalApplicationVersion=\(originalApplicationVersion) ?? ""
...
>
"""
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+        let firstSeen: Date
+        let nonSubscriptions: [String: [[String: Any]]]
+        let entitlements: [String: Any]
+        let nonSubscriptionTransactions: [Transaction]
+        let allPurchases: [String: Any]
+        let expirationDatesByProduct: [String: Date?]
+        let purchaseDatesByProduct: [String: Date?]
+
+        init?(subscriberData: [String: Any]) {
+            let subscriptions = subscriberData["subscriptions"] as? [String: [String: Any]] ?? [String: [String: Any]]()
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:218]: I think you should be able to make the second part of this assignment `?? [:]`, since the compiler should have enough info to infer the type. 
same for others that do the same

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+            // Metadata
+            self.originalApplicationVersion = subscriberData["original_application_version"] as? String ?? nil
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:221]: the `?? nil` isn't needed

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:221]: what _does_ happen in obj-c if we add `nil`? does it translate to `NSNull`? If so, does that match the previous behavior?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions = subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(withSubscriptionsData: nonSubscriptions, dateFormatter: DateFormatter.dateFormatter)
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:249]: we should probably inject the dependency, even if just as a param with a default value in the constructor, so it's easier to test, and so it's easier for us to understand dependencies between our objects

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set<String>(dates.keys.filter({
+            if let nonNullDate = dates[$0] as? Date {
+                return isAfterReferenceDate(date: nonNullDate)
+            }
+            return true
+        }))
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:272]: is the `Set<String>` here needed? 
would doing 
```swift
return dates.keys.filter {
    guard let nonNullDate = dates[$0] as? Date else { return true }
    return isAfterReferenceDate(date: nonNullDate)
}
```
be equivalent?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: Any) -> URL? {
+        if let urlString = url as? String {
+            return URL.init(string: urlString)
+        }
+        return nil
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:281]: `.init` isn't needed, this is equivalent to 
```swift
class func parseURL(url maybeURL: Any) -> URL? {
    guard let urlString = maybeURL as? String else { return }
    return URL(string: urlString)
}
```

also, do we need to have the method accept `Any`?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    class func parseDate(date: Any, withDateFormatter: DateFormatter) -> Date? {
+        if let dateString = date as? String {
+            return DateFormatter.dateFormatter.date(from: dateString)
+        }
+        return nil
+    }
+
+    class func parseExpirationDate(expirationDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: expirationDates, label: "expires_date") }
+
+    class func parsePurchaseDate(purchaseDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: purchaseDates, label: "purchase_date") }
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:293]: we should avoid doing single-liners if the lines go over 120 chars, could you add a line break?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+
+    class func parseDatesIn(dates: [String: Any], label: String) -> [String: Date?] {
+        return dates.mapValues {
+            if let dateString = ($0 as? [String: String] ?? [String: String]())[label] {
+                return DateFormatter.dateFormatter.date(from: dateString)
+            } else {
+                return nil
+            }
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:301]: chaining conditions here might be helpful, i.e.: 
```swift
return dates.mapValues { maybeDatesByKey in
    if let datesByKey = maybeDatesByKey as? [String: String],
       let dateString = datesByKey[label] {
       return DateFormatter.dateFormatter.date(from: dateString)
    }
    return nil
}
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+        }
+    }
+
+}
+
+private extension DateFormatter {
+
+    static let dateFormatter: DateFormatter = {
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:309]: this syntax reads a little weird to me to be honest: 
```Swift
DateFormatter.dateFormatter.foo
```

perhaps this could just be a variable on the class? so it's just `dateFormatter.foo`?

**Review by aboedo — COMMENTED** [review]
great work on this! left a few comments, still reviewing

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    @objc public class func currentSchemaVersion() -> String { "2" }
+
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? { expirationDatesByProduct[forProductIdentifier] ?? nil }
```
[the remark this answers is not in the document]
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:132]: that's what i thought should be the case (for all of these), but when i remove `?? nil` i get this error:
`Value of optional type 'Date??' must be unwrapped to a value of type 'Date?'`. I think since it's possible the dictionary just doesn't contain a value with that key?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 10 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:293]: should lint have caught this? or is that defined somewhere in particular?

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:293]: I think we might have disabled the rule at first since we already had a decent amount of code before adding the linter

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 1 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:162]: i could've sworn i tried and there was a reason we couldn't, but let me try again..

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 8 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:272]: `keys` returns an Array, so it looks like I do have to construct a set from it...but i like your filtering code better 👍

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 9 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:281]: was just me trying to follow what objc did originally, but no, we can just do string since i pull it from the dictionary as a string anyway

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 12 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:309]: agreed it looked weird, was following the `use private extensions of basic types to define constants` in swift sample file, but i suppose this isn't quite the case you were referring to... will update! 👍

Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift]: @aboedo made it a static member. still requires `Self.`, but i think it's a little more clear. let me know your thoughts

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+
+    /// Entitlements attached to this purchaser info
+    @objc public let entitlements: EntitlementInfos
+
+    /// All *subscription* product identifiers with expiration dates in the future.
+    @objc public var activeSubscriptions: Set<String> { activeKeys(dates: expirationDatesByProduct) }
+
+    /// All product identifiers purchases by the user regardless of expiration.
+    @objc public var allPurchasedProductIdentifiers: Set<String> {
+        return self.nonConsumablePurchases.union(expirationDatesByProduct.keys)
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:21]: `pod lib lint` complains about this one and line 121, because we're using `nonConsumablePurchases`, but it's deprecated in favor of `nonSubscriptionTransactions`. I think we should be able to get the same behavior using `nonSubscriptionTransactions`, right?

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:21]: given a `PurchaserInfo` object, would this value ever change? if not, then maybe we can make it a `let` and set the value in the initializer

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+    @objc public let originalApplicationVersion: String?
+
+    // TODO after migration make this internal
+    @objc public let schemaVersion: String?
+
+    private let expirationDatesByProduct: [String: Date?]
+
+    private let purchaseDatesByProduct: [String: Date?]
+
+    private let originalData: [String: Any]
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:89]: [not for this PR] I think we might wanna consider making this kind of thing public, that way new backend features can be supported by older SDK versions by just accessing the originalData

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+
+    private static var dateFormatter: DateFormatter {
+        let formatter = DateFormatter()
+        formatter.timeZone = TimeZone.init(secondsFromGMT: 0)
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:93]: unnecessary `init(`
```swift
formatter.timeZone = TimeZone(secondsFromGMT: 0)
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+                                             purchasesData: subscriberData.allPurchases,
+                                             dateFormatter: Self.dateFormatter,
+                                             requestDate: requestDate)
+
+        self.expirationDatesByProduct = subscriberData.expirationDatesByProduct
+        self.purchaseDatesByProduct = subscriberData.purchaseDatesByProduct
+    }
+
+    // TODO after migration make this internal
+    @objc public class func currentSchemaVersion() -> String { "2" }
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:134]: would making it a `static let` work? 
i.e.: 
```swift
 @objc public static let currentSchemaVersion = "2"
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? {
+        return expirationDatesByProduct[forProductIdentifier] ?? nil
+
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:145]: extra line break

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+    @objc public class func currentSchemaVersion() -> String { "2" }
+
+    /**
+     Get the expiration date for a given product identifier. You should use Entitlements though!
+    
+     @param forProductIdentifier Product identifier for product
+    
+     @return The expiration date for `productIdentifier`, `nil` if product never purchased
+     */
+    @objc public func expirationDate(forProductIdentifier: String) -> Date? {
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:143]: `forProductIdentifier` doesn't really work well as a parameter name. If we want to keep this naming for `objc` compatibility, then we should consider addding a label, i.e.: 

```swift
@objc public func expirationDate(forProductIdentifier productIdentifier: String) -> Date? {
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+     @note This can be nil, see -[RCPurchases restoreTransactionsForAppStore:]
+     */
+    @objc public let originalApplicationVersion: String?
+
+    // TODO after migration make this internal
+    @objc public let schemaVersion: String?
+
+    private let expirationDatesByProduct: [String: Date?]
+
+    private let purchaseDatesByProduct: [String: Date?]
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:87]: perhaps we should be more explicit here with the fact that the keys aren't products, but rather product identifiers? 
i.e.: `expirationDatesByProductId`

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 19 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:143]: same thinking for other methods that have `forFoo` as the param name

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+            guard let nonNullDate = dates[$0] as? Date else { return true }
+            return isAfterReferenceDate(date: nonNullDate)
+        })
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: String) -> URL? { URL(string: url) }
+
+    class func parseDate(dateString: String, withDateFormatter: DateFormatter) -> Date? {
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:304]: the `withDataFormatter` style isn't really needed for swift, this could be 
```swift
class func parseDate(dateString: String, dateFormatter: DateFormatter) -> Date? {
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set(dates.keys.filter {
+            guard let nonNullDate = dates[$0] as? Date else { return true }
+            return isAfterReferenceDate(date: nonNullDate)
+        })
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: String) -> URL? { URL(string: url) }
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:302]: do these need to be static? 
I usually try to avoid static functions because they're harder to mock. With these being private it shouldn't cause issues though.

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:302]: also, I'm kinda wondering whether having this function makes sense... I'm not sure it adds much clarity: 
```swift
self.managementURL = parseURL(url: managementUrlString)
// vs
self.managementURL = URL(string: url)
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 21 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:304]:
```swift
let firstSeenDate = PurchaserInfo.parseDate(dateString: originalPurchaseDateString,
                                            withDateFormatter: PurchaserInfo.dateFormatter)
// vs
let firstSeenDate = dateFormatter.date(from: originalPurchaseDateString)
```

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:304]: I'm now realizing that these are probably static because we have a static date formatter.
And formatters are costly to build, so we really only want to have one, so that's why we have it as static. 

I think the best solution for a situation like this one is to have the formatter be injected into the `init`, and have a singleton instance living somewhere else. 

I believe you had it like that and I made you change it 🤦 That's entirely my bad. I think what had seemed odd to me was the syntax `DateFormatter.dateFormatter.foo`. 

However, if we had something like 
- `DateFormatter.iso8601Formatter` as a static property in an extension to `DateFormatter`, ideally outside this file
- had a property in PurchaserInfo `private let dateFormatter: DateFormatter`
- send it in the `init`, i.e.: 
```swift
init?(subscriberData: [String: Any], dateFormatter: DateFormatter = .iso8601DateFormatter()) {
    self.dateFormatter = dateFormatter
    ...
```
then we can use it as 

```swift
let firstSeenDate = dateFormatter.date(from: originalPurchaseDateString)
```

and, should we ever need to test that the formatting is done appropriately, we can always mock the formatter here, and we even have quick access to writing tests for the `iso8601Formatter` to make sure that we're creating it correctly.

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:304]: after writing this it dawned on me that we already have a date formatter built in, for the local receipt parser, so we can reuse [that one](https://github.com/RevenueCat/purchases-ios/blob/main/PurchasesCoreSwift/LocalReceiptParsing/DataConverters/ISO3601DateFormatter.swift)
I wish I'd thought to name it `iso8601DateFormatter` back then, though.

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 22 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:302]: agreed on removing that function... parse date probably isn't much of a help either...

Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:302]: as for your first comment..they didn't NEED to be, just thought in general we should make things static if we can?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 14 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:21]: sure, seems safe enough...

on the second comment -- i guess computed properties are computed for every `get`, but thoughts on making it a lazy variable instead? could probably do that with `latestExpirationDate` and `activeSubscriptions` too...or is there some advantage to having it set in init?

Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:21]: for `activeSubscriptions` and others that compare against the current date, we have to keep them as computed properties since there's a chance that you'd get a different result when calling it at different times

Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:21]: 🤦 yep good catch

**Comment by beylmk** [comment]
@aboedo i believe i addressed all your comments! let me know if you'd prefer the dateformatter stuff to just be merged along with this..

File: PublicSDKAPITester/RCPurchasesAPI.m [hunk 23]
```diff
@@ -64,7 +64,7 @@ + (void)checkAPI {
     appUserID = [p appUserID];
     isAnonymous = [p isAnonymous];
     
-    RCPurchaserInfo *pi = [[RCPurchaserInfo alloc] init];
+    RCPurchaserInfo *pi = [[RCPurchaserInfo alloc] initWithData: [[NSDictionary alloc] init]];
```
Comment (aboedo) [inline PublicSDKAPITester/RCPurchasesAPI.m:67]: extra space after `:`

You can also do this a little more simply by using a dict literal, i.e.: 
```objective-c
RCPurchaserInfo *pi = [[RCPurchaserInfo alloc] initWithData:@{}];
```

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 16 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:93]: I believe this gets addressed in a follow-up pr, right?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 4 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:218]: this one is still valid

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 16 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:93]: yes!

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:196]: > also, do we need these to be NSDictionaries?
> would doing something `selfJson == otherJson` not work? is it because of the `AnyHashable` thing?

yes, because the keys are `Any` which doesn't conform to equatable.


> should this be `NSDictionary(dictionary: selfJson).isEqual(to: NSDictionary(dictionary: otherJson))`?

that version of isEqual takes [AnyHashable: Any], but making a dictionary feels safer/clearer to me so i'll update it anyway

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 7 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift]: @aboedo cool with me adding this in the dateformatter pr? since it has the convenience init

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 6 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:242]: previous behavior would set originalApplicationVersion to `nil` if `data["original_application_version"]` was of type `NSNull`.

and good question, i'll confirm..

Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:242]: @aboedo previously, we'd set `originalApplicationVersion` to `nil` if `data["original_application_version"] was type NSNull... 

it looks like whether we add nil or not in this line, it'll translate to nil in objc. so we should be good

**Comment by beylmk** [comment]
@aboedo should be ready again!

File: Purchases/Public/RCPurchases.h [hunk 24]
```diff
@@ -101,7 +101,7 @@ NS_SWIFT_NAME(Purchases)
 /**
  Configures an instance of the Purchases SDK with a specified API key. The instance will be set as a singleton. You should access the singleton instance using [RCPurchases sharedPurchases]
 
- @note Use this initializer if your app does not have an account system. `RCPurchases` will generate a unique identifier for the current device and persist it to `NSUserDefaults`. This also affects the behavior of `restoreTransactionsForAppStoreAccount`.
+ @note Use this initializer if your app does not have an account system. `RCPurchases` will generate a unique identifier for the current device and persist it to `NSUserDefaults`. This also affects the behavior of `restoreTransactionsWithCompletionBlock`.
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.h:104]: while we're at it, could you add line breaks so that this doesn't scroll horizontally? I've been using ~120 chars as the limit

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 25]
```diff
@@ -0,0 +1,315 @@
...
+     If there are no active subscriptions it will be null.
+     If there are multiple for different platforms, it will point to the App Store
+     */
+    @objc public let managementURL: URL?
+
+    /**
+    Returns the purchase date for the version of the application when the user bought the app.
+    Use this for grandfathering users when migrating to subscriptions.
+
+    Note: This can be `nil`, see `RCPurchases restoreTransactionsWithCompletionBlock`
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:69]: https://github.com/realm/jazzy#supported-documentation-keywords

I think this would have to be `[RCPurchases restoreTransactionsWithCompletionBlock:]`, but then again, since we're migrating, perhaps we should use the swift syntax, i.e.: `Purchases.restoreTransactions(completionBlock:)`

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 26]
```diff
@@ -0,0 +1,315 @@
...
+     URL to manage the active subscription of the user.
+     If this user has an active iOS subscription, this will point to the App Store,
+     if the user has an active Play Store subscription it will point there.
+     If there are no active subscriptions it will be null.
+     If there are multiple for different platforms, it will point to the App Store
+     */
+    @objc public let managementURL: URL?
+
+    /**
+    Returns the purchase date for the version of the application when the user bought the app.
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:66]: these should be `- returns:`

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 2 above)
Comment (aboedo) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:11]: did you get jazzy to work? lmk if you need a hand

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 27]
```diff
@@ -0,0 +1,315 @@
...
+
+    private static var dateFormatter: DateFormatter {
+        let formatter = DateFormatter()
+        formatter.timeZone = TimeZone(secondsFromGMT: 0)
+        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
+        formatter.locale = Locale(identifier: "en_US_POSIX")
+        return formatter
+    }
+
+    @objc public init?(data: [String: Any]) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:100]: let's add a TODO make internal here

File: PurchasesCoreSwift/Public/PurchaserInfo.swift [hunk 28]
```diff
@@ -0,0 +1,315 @@
...
+    @objc public func purchaseDate(forEntitlement entitlementIdentifier: String) -> Date? {
+        return entitlements[entitlementIdentifier]?.latestPurchaseDate
+    }
+
+    // TODO after migration make this internal and remove objc rename
+    @objc(JSONObject) public func jsonObject() -> [String: Any] {
+        return originalData.merging(
+            ["schema_version": PurchaserInfo.currentSchemaVersion],
+            uniquingKeysWith: { (current, _) in current })
+    }
```
Comment (aboedo) [inline PurchasesCoreSwift/Public/PurchaserInfo.swift:185]: perhaps we can keep all of these will-be-internal _after_ the public ones? that way we're following the public first, internal second, private last structure

**Review by aboedo — COMMENTED** [review]
looking great, are you looking to do the documentation updates on this PR or a separate one?

File: PurchasesCoreSwift/Public/PurchaserInfo.swift (hunk 15 above)
Comment (beylmk) [reply PurchasesCoreSwift/Public/PurchaserInfo.swift:90]: opened an issue for the idea: https://github.com/RevenueCat/purchases-ios/issues/708

**Review by aboedo — APPROVED** [review]
